### PR TITLE
add rasterio env context for STAC

### LINKED ISSUE #550: make use of file stac-extension `header_size` in MultiBaseReader
in https://github.com/stac-extensions/file there is a `header_size` info which in theory could be used to increase `GDAL_INGESTED_BYTES_AT_OPEN` https://developmentseed.org/titiler/advanced/performance_tuning/#gdal_ingested_bytes_at_open and reduce the number of GET request for accessing the asset.

## How to 

- refactor `MultiBaseReader._get_asset_url` 
https://github.com/cogeotiff/rio-tiler/blob/da7b8b315f53418d6e8b36138f6131020da5e484/rio_tiler/io/base.py#L270-L273

to return a dictionary in form of `{"url": str, "env": Optional[Dict]}`

- The `env` could then be used in https://github.com/cogeotiff/rio-tiler/blob/da7b8b315f53418d6e8b36138f6131020da5e484/rio_tiler/io/base.py#L304-L307

like 
```python
def _reader(asset: str, **kwargs: Any) -> Dict:
    asset_info = self._get_asset_url(asset)
    with rasterio.Env(**asset_info.env):
        with self.reader(asset_info.url, tms=self.tms, **self.reader_options) as cog:  # type: ignore
            return cog.info()
```

cc @geospatial-jeff

### PR DESCRIPTION
**Pull request by vincentsarago**
closes #550

### CODE REVIEW

File: rio_tiler/io/base.py [hunk 1]
```diff
@@ -302,9 +305,12 @@ def info(
             assets = (assets,)
 
         def _reader(asset: str, **kwargs: Any) -> Dict:
-            url = self._get_asset_url(asset)
-            with self.reader(url, tms=self.tms, **self.reader_options) as src:  # type: ignore
-                return src.info()
+            asset_meta = self._get_asset_info(asset)
+            print(asset_meta)
```
Comment (vincentsarago) [inline rio_tiler/io/base.py:309]:
```suggestion
            print(asset_meta)
```

Comment (vincentsarago) [inline rio_tiler/io/base.py:309]:
```suggestion
```

File: rio_tiler/io/base.py [hunk 2]
```diff
@@ -259,6 +260,8 @@ class MultiBaseReader(SpatialMixin, metaclass=abc.ABCMeta):
 
     assets: Sequence[str] = attr.ib(init=False)
 
+    ctx: Any = attr.ib(init=False, default=contextlib.nullcontext)
```
Comment (vincentsarago) [inline rio_tiler/io/base.py:263]: in practice `ctx` will be `rasterio.Env` but this is not mandatory and could be let to `contextlib.nullcontext` is the MultiBaseReader class do not use a rasterio based reader

File: rio_tiler/io/stac.py [hunk 3]
```diff
@@ -224,4 +229,11 @@ def _get_asset_url(self, asset: str) -> str:
         if asset not in self.assets:
             raise InvalidAssetName(f"{asset} is not valid")
 
-        return self.item.assets[asset].get_absolute_href()
+        asset_info = self.item.assets[asset]
+        info = AssetInfo(url=asset_info.get_absolute_href())
+
+        if "file:header_size" in asset_info.extra_fields:
+            h = asset_info.extra_fields["file:header_size"]
+            info["env"] = {"GDAL_INGESTED_BYTES_AT_OPEN": 16384 * math.ceil(h / 16384)}
```
Comment (vincentsarago) [inline rio_tiler/io/stac.py:237]: ✨

**Comment by vincentsarago** [comment]
@geospatial-jeff any feedback on this?

File: rio_tiler/io/stac.py (hunk 3 above)
Comment (geospatial-jeff) [reply rio_tiler/io/stac.py:237]: Isn't this the same as `{"GDAL_INGESTED_BYTES_AT_OPEN": h}`

**Review by geospatial-jeff — APPROVED** [review]

File: rio_tiler/io/stac.py (hunk 3 above)
Comment (vincentsarago) [reply rio_tiler/io/stac.py:237]: I guess it should

**Pull request merged**